### PR TITLE
avatar changeable, profile connected with API request (fixes: #111)(fixes: #120) 

### PR DESCRIPTION
**Pull request by Hjortsberg**
started work on the avatar controller and model on the
server implementation.

### CODE REVIEW

File: controller/avatar/avatar.go [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+	}
+
+	username, ok := s.Values["username"].(string)
+	if !ok || username == "" {
+		http.Error(w, http.StatusText(http.StatusNoContent), http.StatusNoContent)
+		return
+	}
+    mediaType, params, err := mime.ParseMediaType(r.Header.Get("Content-Type"))
+    if err != nil {
+        http.Error(w, "expected a avatar multipart, recieved something else", http.StatusBadRequest)
```
Comment (willeponken) [inline controller/avatar/avatar.go]: an

**Comment by Hjortsberg** [comment]
@willeponken ready for review! 😈

File: controller/avatar/avatar.go (hunk 1 above)
Comment (Hjortsberg) [reply controller/avatar/avatar.go]: that
was
fast

File: www/.eslintrc.json [hunk 2]
```diff
@@ -7,6 +7,7 @@
   },
   "env": { "es6": true },
   "rules": {
-    "require-jsdoc": "off"
+    "require-jsdoc": "off",
+    "no-unused-vars": "error"
```
[the remark this answers is not in the document]
Comment (Hjortsberg) [reply www/.eslintrc.json]: I did, there seems to be configuration problem that we need to solve before we can use that.

File: model/avatar.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        `, username).Scan(&avatar.Username, &avatar.Avatar)
+    return
+}
+
+func randomGender() govatar.Gender {
+    r := rand.Intn(1)
+    if r == 0 {
+        return govatar.MALE
+    }
+    return govatar.FEMALE
```
[the remark this answers is not in the document]
Comment (Hjortsberg) [reply model/avatar.go]: 🙄

[the remark this answers is not in the document]
Comment (Hjortsberg) [reply model/avatar.go]: I'm a bit unsure about our random function here, the population is like 100% males right now.

[the remark this answers is not in the document]
Comment (Hjortsberg) [reply model/avatar.go]: Fixed now! Gotta know your math notations!

File: www/.eslintrc.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (willeponken) [reply www/.eslintrc.json]: OKAY I SEE THANK YOU COME AGAIN

[the remark this answers is not in the document]
Comment (willeponken) [reply www/.eslintrc.json]: PLZ***

File: controller/avatar/avatar.go [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+	"github.com/gorilla/mux"
+	"github.com/lib/pq"
+)
+
+func createAvatar(w http.ResponseWriter, r *http.Request) {
+	var a model.Avatar
+
+	s, err := session.Get(r, "roaster_auth")
+	if err != nil {
+		log.Println(err)
```
Comment (willeponken) [inline controller/avatar/avatar.go]: Should return Not Authenticated (402?)

File: controller/avatar/avatar.go [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+	}
+
+	username, ok := s.Values["username"].(string)
+	if !ok || username == "" {
+		http.Error(w, http.StatusText(http.StatusNoContent), http.StatusNoContent)
```
Comment (willeponken) [inline controller/avatar/avatar.go]: Should return Not Authenticated (402?)

File: controller/avatar/avatar.go [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+		return
+	}
+	mediaType, params, err := mime.ParseMediaType(r.Header.Get("Content-Type"))
+	if err != nil {
+		http.Error(w, "expected an avatar multipart, recieved something else", http.StatusBadRequest)
```
Comment (willeponken) [inline controller/avatar/avatar.go]: RECEIVED

File: controller/avatar/avatar.go [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+				http.Error(w, "too large file (greater than 10MB)", http.StatusRequestEntityTooLarge)
+			}
+			err = model.PutAvatar(a)
+			if err != nil {
+				log.Println(err)
+				http.Error(w, "error occured when trying to save the image", http.StatusInternalServerError)
+			}
+		}
+	} else {
+		http.Error(w, "received wrong Content-Type, expected multipart/*", http.StatusBadRequest)
```
Comment (willeponken) [inline controller/avatar/avatar.go]: Great spelling!

File: controller/avatar/avatar.go [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+			if err != nil {
+				http.Error(w, "error while parsing multipart message", http.StatusBadRequest)
+				return
+			}
+			a.Avatar, err = ioutil.ReadAll(p)
+			if err != nil {
+				http.Error(w, "error while parsing all in multipart message", http.StatusBadRequest)
+			}
+			a.Username = username
+			if len(a.Avatar) >= 10*1000*1000 {
```
Comment (willeponken) [inline controller/avatar/avatar.go]: DRY!!!!

File: controller/avatar/avatar.go [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+		for {
+			p, err := mr.NextPart()
+			if err == io.EOF {
+				return
+			}
+			if err != nil {
+				http.Error(w, "error while parsing multipart message", http.StatusBadRequest)
+				return
+			}
+			a.Avatar, err = ioutil.ReadAll(p)
```
Comment (willeponken) [inline controller/avatar/avatar.go]: DRY!

File: controller/avatar/avatar.go [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+			if err != nil {
+				http.Error(w, "error while parsing all in multipart message", http.StatusBadRequest)
+			}
+			a.Username = username
+			if len(a.Avatar) >= 10*1000*1000 {
+				http.Error(w, "too large file (greater than 10MB)", http.StatusRequestEntityTooLarge)
```
Comment (willeponken) [inline controller/avatar/avatar.go]: Missing `return`

File: controller/avatar/avatar.go [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+			}
+			err = model.PutAvatar(a)
+			if err != nil {
+				log.Println(err)
+				http.Error(w, "error occured when trying to save the image", http.StatusInternalServerError)
```
Comment (willeponken) [inline controller/avatar/avatar.go]: Missing `return`

File: controller/avatar/avatar.go (hunk 7 above)
Comment (willeponken) [reply controller/avatar/avatar.go]: Missing `return`

File: controller/avatar/avatar.go [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+			if err == io.EOF {
+				return
+			}
+			if err != nil {
+				http.Error(w, "error while parsing multipart message", http.StatusBadRequest)
+				return
+			}
+			a.Avatar, err = ioutil.ReadAll(p)
+			if err != nil {
+				http.Error(w, "error while parsing all in multipart message", http.StatusBadRequest)
```
Comment (willeponken) [inline controller/avatar/avatar.go]: Missing `return`

File: controller/avatar/avatar.go (hunk 4 above)
Comment (willeponken) [reply controller/avatar/avatar.go]: Missing `return`

File: model/avatar.go [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+func PutAvatar(avatar Avatar) (err error) {
+	_, err = database.Exec(`
+        INSERT INTO "avatar"
+        (username, avatar)
+        VALUES
+        (TRIM($1), $2)
+        ON CONFLICT
+        (username)
+        DO UPDATE SET 
+        avatar = EXCLUDED.avatar
```
Comment (willeponken) [inline model/avatar.go]: NICE!!!!!

File: model/avatar.go [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+    `, avatar.Username, avatar.Avatar)
+	return
+}
+
+// EncodeToPNG encodes a Go image.Image to a PNG.
+func encodeToPNG(img image.Image) (raw []byte, err error) {
+	buf := new(bytes.Buffer)
+	err = png.Encode(buf, img)
+	raw = buf.Bytes()
+	return
```
Comment (willeponken) [inline model/avatar.go]: DRY!

File: model/avatar.go [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+	err = database.QueryRow(`
+        SELECT username, avatar
+        FROM "avatar"
+        WHERE username=TRIM($1)
+        `, username).Scan(&avatar.Username, &avatar.Avatar)
+	return
+}
+
+func randomGender() govatar.Gender {
+	r := rand.Intn(2)
```
Comment (willeponken) [inline model/avatar.go]: Great fix!!!!

**Pull request merged**